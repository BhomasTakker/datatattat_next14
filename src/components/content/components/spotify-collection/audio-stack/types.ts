import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
