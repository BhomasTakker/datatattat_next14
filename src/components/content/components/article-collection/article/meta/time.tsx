import { StyleSheet } from "@/types/css";

type Time = {
	time: Date | string;
	styles: StyleSheet;
};

const pluralise = (unit: number, str: string) => {
	return unit > 1 ? `${unit} ${str}s` : `${unit} ${str}`;
};

// Update to use dayjs fully?
// Article time
export const Time = ({ time, styles }: Time) => {
	// cretae a class to get this more easily etc
	const date = new Date(time);
	const now = new Date(Date.now());
	const hours = date.getHours();
	const minutes = date.getMinutes();

	const useHours = hours < 10 ? `0${hours}` : hours;
	const useMinutes = minutes < 10 ? `0${minutes}` : minutes;

	// @ts-expect-error - arithmatic on date - seems fine?
	const diffTime = Math.abs(now - date);
	const diffSeconds = Math.floor(diffTime / 1000);
	const diffMinutes = Math.floor(diffTime / (1000 * 60));
	const diffHours = Math.floor(diffTime / (1000 * 60 * 60));
	const diffDays = Math.floor(diffTime / (1000 * 60 * 60 * 24));
	const diffWeeks = Math.floor(diffTime / (1000 * 60 * 60 * 24 * 7));
	const diffRoughMonths = Math.floor(diffTime / (1000 * 60 * 60 * 24 * 30));
	// const diffYears = Math.floor(diffTime / (1000 * 60 * 60 * 24 * 365));

	let displayTime = "";

	// Quick change seems to suit
	// reformat this all and look at
	// Intl.DateTimeFormat
	switch (true) {
		case !!diffRoughMonths:
			displayTime = `${date.getDate()} ${date.toLocaleDateString("en-GB", {
				month: "short",
			})} ${date.getFullYear()}`;
			break;
		case !!diffWeeks:
			displayTime = `${date.getDate()} ${date.toLocaleDateString("en-GB", {
				month: "short",
			})} ${useHours}:${useMinutes}`;
			break;
		case !!diffDays:
			displayTime = `${new Intl.DateTimeFormat("en-US", {
				weekday: "short",
			}).format(date)} ${useHours}:${useMinutes}`;
			break;
		case !!diffHours:
			displayTime = `${pluralise(diffHours, "hour")} ago`;
			break;
		case !!diffMinutes:
			displayTime = `${pluralise(diffMinutes, "minute")} ago`;
			break;
		case !!diffSeconds:
			displayTime = "Now!";
			break;
		default:
			displayTime = `${date.getDate()} ${date.toLocaleDateString("en-GB", {
				month: "short",
			})} ${date.getFullYear()}`;
	}

	return (
		<time className={styles.time} dateTime={date.toISOString()}>
			{displayTime}
		</time>
	);
};
