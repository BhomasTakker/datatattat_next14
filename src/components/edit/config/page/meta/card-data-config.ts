import { ShowInputProps } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../../inputs/inputs";

export const CARD_DATA: ShowInputProps = {
	id: "showCardData",
	type: EditInputs.show,
	label: "Create Card Data",
	inputs: [
		{
			id: "cardData",
			type: EditInputs.inputList,
			label: "Card Data",
			inputs: [
				// {
				// 	id: "twitter:card",
				// 	type: EditInputs.text,
				// 	label: "Card",
				// },
				// {
				// 	id: "twitter:creator",
				// 	type: EditInputs.text,
				// 	label: "Creator",
				// },
				{
					id: "description",
					type: EditInputs.text,
					label: "Description",
				},
				{
					id: "image",
					type: EditInputs.text,
					label: "Image",
				},
				{
					id: "image:alt",
					type: EditInputs.text,
					label: "Image Alt",
				},
				{
					id: "locale",
					type: EditInputs.text,
					label: "Locale",
				},
				{
					id: "site_name",
					type: EditInputs.text,
					label: "Site Name",
				},
				{
					id: "title",
					type: EditInputs.text,
					label: "Title",
				},
				// {
				// 	id: "og:type",
				// 	type: EditInputs.text,
				// 	label: "Type",
				// },
				{
					id: "url",
					type: EditInputs.text,
					label: "Url",
				},
			],
		},
	],
};
