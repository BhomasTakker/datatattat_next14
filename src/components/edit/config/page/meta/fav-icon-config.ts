import { ArrayInputProps } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../../inputs/inputs";

enum typeOptions {
	PNG = "image/png",
	JPEG = "image/jpeg",
}

enum relOptions {
	ICON = "icon",
	APPLE = "apple-touch-icon-precomposed",
}

enum sizes {
	ANY = "any",
	_16x16 = "16x16",
	_32x32 = "32x32",
	_36x36 = "36x36",
	_48x48 = "48x48",
	_57x57 = "57x57",
	_60x60 = "60x60",
	_72x72 = "72x72",
	_76x76 = "76x76",
	_96x96 = "96x96",
	_114x114 = "114x114",
	_120x120 = "120x120",
	_128x128 = "128x128",
	_144x144 = "144x144",
	_152x152 = "152x152",
	_180x180 = "180x180",
	_192x192 = "192x192",
	_384x384 = "384x384",
	_512x512 = "512x512",
}

export const FAV_ICON: ArrayInputProps = {
	id: "favIcons",
	type: EditInputs.array,
	label: "FavIcons",
	title: "Add Fav Icon",
	input: {
		id: "favIcon",
		type: EditInputs.inputList,
		label: "Fav Icon",
		inputs: [
			{
				id: "rel",
				type: EditInputs.select,
				label: "Rel",
				defaultValue: relOptions.ICON,
				required: true,
				options: [relOptions.ICON, relOptions.APPLE],
			},
			{
				id: "type",
				type: EditInputs.select,
				label: "Source type",
				defaultValue: typeOptions.PNG,
				required: true,
				options: [typeOptions.PNG, typeOptions.JPEG],
			},
			{
				id: "href",
				type: EditInputs.text,
				label: "Source",
				required: true,
			},
			{
				id: "sizes",
				type: EditInputs.select,
				label: "Size",
				defaultValue: sizes.ANY,
				required: true,
				options: [
					sizes.ANY,
					sizes._16x16,
					sizes._32x32,
					sizes._36x36,
					sizes._48x48,
					sizes._57x57,
					sizes._60x60,
					sizes._72x72,
					sizes._76x76,
					sizes._96x96,
					sizes._114x114,
					sizes._120x120,
					sizes._128x128,
					sizes._144x144,
					sizes._152x152,
					sizes._180x180,
					sizes._192x192,
					sizes._384x384,
					sizes._512x512,
				],
			},
		],
	},
};
