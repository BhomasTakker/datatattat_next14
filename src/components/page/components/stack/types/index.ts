import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
