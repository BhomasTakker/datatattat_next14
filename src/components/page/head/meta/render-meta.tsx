import { UnknownObject } from "@/types/utils";

interface Meta {
	data: UnknownObject;
}

export const renderMeta = ({ data }: Meta) => {
	const metadata = [];
	for (const key in data) {
		const value = data[key] as string;
		metadata.push(
			<meta
				key={key}
				data-rh="true"
				name={key}
				property={key}
				content={value}
			></meta>
		);
	}

	return [...metadata];
};
