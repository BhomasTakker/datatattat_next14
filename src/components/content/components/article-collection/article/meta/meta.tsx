import { StyleSheet } from "@/types/css";
import { Time } from "./time";

type Props = {
	styles: StyleSheet;
	publisher?: string;
	published?: Date | string;
};

export const Meta = ({ published, publisher, styles }: Props) => {
	return (
		<div className={styles.meta}>
			{publisher ? (
				<p className={styles.publishers} data-testid="publisher">
					{publisher}
				</p>
			) : (
				<></>
			)}
			{published ? <Time styles={styles} time={published} /> : null}
		</div>
	);
};
