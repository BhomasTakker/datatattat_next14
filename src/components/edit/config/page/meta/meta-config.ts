import { InputListProps } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../../inputs/inputs";
import { CARD_DATA } from "./card-data-config";
import { FAV_ICON } from "./fav-icon-config";

export const META_CONFIG: InputListProps = {
	id: "meta",
	type: EditInputs.inputList,
	label: "Meta Object",

	inputs: [
		{
			id: "metaPropertiesTitle",
			type: EditInputs.title,
			title: "Meta Properties",
		},

		{
			id: "pageTitle",
			type: EditInputs.text,
			label: "Page Title",
		},
		{
			id: "pageDescription",
			type: EditInputs.text,
			label: "Page Description",
		},
		{
			id: "pageKeywords",
			type: EditInputs.text,
			label: "Page Keywords",
		},
		{
			id: "pageImage",
			type: EditInputs.text,
			label: "Page Image",
		},
		FAV_ICON,
		CARD_DATA,
		// Should be created by the system
		// {
		// 	id: "pageType",
		// 	type: EditInputs.text,
		// 	label: "Page Type",
		// },
		// {
		// 	id: "pageUrl",
		// 	type: EditInputs.text,
		// 	label: "Page URL",
		// },
		// {
		// 	id: "pageAuthor",
		// 	type: EditInputs.text,
		// 	label: "Page Author",
		// },
		// {
		// 	id: "pagePublished",
		// 	type: EditInputs.text,
		// 	label: "Page Published",
		// },
		// {
		// 	id: "pageModified",
		// 	type: EditInputs.text,
		// 	label: "Page Modified",
		// },
	],
};
