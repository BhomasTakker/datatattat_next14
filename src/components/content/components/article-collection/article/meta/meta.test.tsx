import React from "react";
import { render, screen } from "@testing-library/react";
import { Meta } from "./meta";
import { StyleSheet } from "@/types/css";
import { Details } from "@/types/data-structures/collection/base";

const mockStyles: StyleSheet = {
	meta: "meta",
	categories: "categories",
	authors: "authors",
	publishers: "publishers",
};

const mockDetails: Details = {
	categories: ["Category1"],
	authors: ["Author1"],
	published: "2023-10-01T00:00:00Z",
	publishers: ["Publisher1"],
};

describe("Meta Component", () => {
	// it("renders categories when provided", () => {
	// 	render(<Meta styles={mockStyles} {...mockDetails} />);
	// 	const categories = screen.queryByText("Category1");

	// 	expect(categories).toBeInTheDocument();
	// 	expect(categories).toBeInTheDocument();
	// });

	// it("renders authors when provided", () => {
	// 	render(<Meta styles={mockStyles} {...mockDetails} />);
	// 	const authors = screen.queryByText("Author1");
	// 	expect(authors).toBeInTheDocument();
	// });

	it("renders publishers when provided", () => {
		render(<Meta styles={mockStyles} {...mockDetails} />);
		const publishers = screen.queryByText("Publisher1");
		expect(publishers).toBeInTheDocument();
	});

	it("renders expected time when provided", () => {
		render(<Meta styles={mockStyles} {...mockDetails} />);
		const published = screen.getByText("1 Oct 2023");
		expect(published).toBeInTheDocument();
	});

	// it("does not render categories when not provided", () => {
	// 	render(
	// 		<Meta styles={mockStyles} {...{ ...mockDetails, categories: [] }} />
	// 	);
	// 	// better test needed
	// 	const categories = screen.queryByText("Category1");
	// 	expect(categories).not.toBeInTheDocument();
	// });

	// it("does not render authors when not provided", () => {
	// 	render(<Meta styles={mockStyles} {...{ ...mockDetails, authors: [] }} />);
	// 	const authors = screen.queryByText("Author1");
	// 	expect(authors).not.toBeInTheDocument();
	// });

	it("does not render publishers when not provided", () => {
		render(
			<Meta styles={mockStyles} {...{ ...mockDetails, publishers: [] }} />
		);
		const publishers = screen.queryByText("Publisher1");
		expect(publishers).not.toBeInTheDocument();
	});

	it("does not render published time when not provided", () => {
		render(<Meta styles={mockStyles} {...{ ...mockDetails, published: "" }} />);
		const published = screen.queryByText("2023-10-01T00:00:00Z");
		expect(published).not.toBeInTheDocument();
	});

	it("renders correctly with all props", () => {
		const { container } = render(<Meta styles={mockStyles} {...mockDetails} />);
		expect(container).toMatchSnapshot();
	});
});
