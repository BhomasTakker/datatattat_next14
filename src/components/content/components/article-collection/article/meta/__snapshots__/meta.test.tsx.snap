// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meta Component renders correctly with all props 1`] = `
<div>
  <div
    class="meta"
  >
    <p
      class="publishers"
      data-testid="publisher"
    >
      Publisher1
    </p>
    <time
      datetime="2023-10-01T00:00:00.000Z"
    >
      1 Oct 2023
    </time>
  </div>
</div>
`;
